decentraland/unity-explorer
# fix: 100avatars-garden signed fetch

## What does this PR change?

Fixes #1473 

There were problems on solving the metadata included in the signed fetch operation:
- `metadata.sceneId`: removed unnecessary prefix so only the scene hash is included
- `metadata.realm.hostname`: reworked how the hostname is resolved for the current realm. When the /about data of the realm has missing comms, then its considered as `main` realm, so the hostname is set to `realm-provider.decentraland.org`. If the realm is a world, then is resolved as: `worlds-content-server.decentraland.org/world/{ensName}`. Otherwise is the host name calculated through the desired realm domain.

The expiration date format in the auth chain now includes milliseconds as it could be required for date parsing on external servers.
The current realm information we were providing to the SDK was wrong, specially the `domain` field. We were providing the realm base URL, but it is needed to provide the domain only, without the path as it is implemented in kernel: https://github.com/decentraland/unity-renderer/blob/4a2f0ecba4235cf481eebbfccc7d664fea96fc4d/browser-interface/packages/shared/apis/host/EnvironmentAPI.ts#L108. This is necessary to properly form the request to claim the wearable to the rewards api:
```
curl 'https://rewards.decentraland.org/api/campaigns/c88873d0-b65d-4b48-bd0d-434d9449e05b/rewards' -X POST
  -H 'Content-Type: application/json'
  -H 'Origin: https://decentraland.org'
  -H 'Referer: https://decentraland.org'
  -H 'x-identity-auth-chain-0: {"type":"SIGNER","payload":"0x86b04fbfccae6abfad341b8f6236a86f0a7e3753","signature":""}'
  -H 'x-identity-auth-chain-1: {"type":"ECDSA_EPHEMERAL","payload":"Decentraland Login\nEphemeral address: 0xC1F07E11FDD9d6636237Ae03A9c3f46AcfBF7745\nExpiration: 2024-08-07T16:01:24.545Z","signature":"0x..."}'
  -H 'x-identity-auth-chain-2: {"type":"ECDSA_SIGNED_ENTITY","payload":"post:/api/campaigns/c88873d0-b65d-4b48-bd0d-434d9449e05b/rewards:1722441919214:{\"origin\":\"https://decentraland.org\",\"sceneid\":\"bafkreic5dxyxeqqt63znrsw6burxjxkvqod3y74tmhleorztjn37miqja4\",\"parcel\":\"-55,-121\",\"tld\":\"org\",\"network\":\"mainnet\",\"isguest\":false,\"realm\":{\"hostname\":\"realm-provider.decentraland.org\",\"protocol\":\"v3\",\"servername\":\"main\"},\"signer\":\"decentraland-kernel-scene\"}","signature":"0x..."}'
  -H 'x-identity-timestamp: 1722441919214'
  -H 'x-identity-metadata: {"origin":"https://decentraland.org","sceneId":"bafkreic5dxyxeqqt63znrsw6burxjxkvqod3y74tmhleorztjn37miqja4","parcel":"-55,-121","tld":"org","network":"mainnet","isGuest":false,"realm":{"hostname":"realm-provider.decentraland.org","protocol":"v3","serverName":"main"},"signer":"decentraland-kernel-scene"}'
  --data-raw '{"campaign_key":"fjg9fbNcSO+R9hn57RiXM8iIc9C2XUtIvQ1DTZRJ4Fs=.w1rItMmAZPh486pOe5bFOjp3jIxgksu/NfRp/M9AWm0=","catalyst":"https://realm-provider.decentraland.org","beneficiary":"0x86b04fbfccae6abfad341b8f6236a86f0a7e3753"}'
```

This is finally how the dcl login request looks:
```
curl 'https://v2.lowpolyhub.com:3000/loginWithDcl' -X POST
  -H 'Referer: https://decentraland.org/'
  -H 'x-identity-auth-chain-0: {"type":"SIGNER","payload":"0x43d803d72ba0a4a785f0e46d3b2366cd224e49df","signature":""}'
  -H 'x-identity-auth-chain-1: {"type":"ECDSA_EPHEMERAL","payload":"Decentraland Login\nEphemeral address: 0x08e9E6684fbcC07218742F949b4e001B99F691f3\nExpiration: 2024-07-31T14:53:29.435Z","signature":"0x..."}'
  -H 'x-identity-auth-chain-2: {"type":"ECDSA_SIGNED_ENTITY","payload":"post:/loginwithdcl:1721833159404:{\"origin\":\"https://decentraland.org\",\"sceneid\":\"bafkreic5dxyxeqqt63znrsw6burxjxkvqod3y74tmhleorztjn37miqja4\",\"parcel\":\"-55,-121\",\"tld\":\"org\",\"network\":\"mainnet\",\"isguest\":false,\"realm\":{\"hostname\":\"realm-provider.decentraland.org\",\"protocol\":\"v3\",\"servername\":\"main\"},\"signer\":\"decentraland-kernel-scene\"}","signature":"0x..."}'
  -H 'x-identity-timestamp: 1721759308313'
  -H 'x-identity-metadata: {"origin":"https://decentraland.org","sceneId":"bafkreic5dxyxeqqt63znrsw6burxjxkvqod3y74tmhleorztjn37miqja4","parcel":"-55,-121","tld":"org","network":"mainnet","isGuest":false,"realm":{"hostname":"realm-provider.decentraland.org","protocol":"v3","serverName":"main"},"signer":"decentraland-kernel-scene"}'
  -H 'Content-Type: application/json'
  -H 'Origin: https://decentraland.org'
  --data-raw '{"wallet":"0x43d803d72ba0a4a785f0e46d3b2366cd224e49df","username":"Lorux0r"}'
```

## How to test the changes?

1. Follow issue steps
2. You should be able to claim the wearable
3. Go to a world, ie: `metadyne.dcl.eth` and check signed fetch is correctly working
4. Also on goerli there are several scenes that uses signed fetch. All of them should keep working:
- 82,-6
- 78,1
- 83,-4
- 81,-9

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
